### PR TITLE
SWATCH-2162: Send metrics using backpressure mecanisms into Kafka

### PR DESCRIPTION
**Pull request by Sgitario**
Jira issue: [SWATCH-2162](https://issues.redhat.com/browse/SWATCH-2162)

## Description
This solution has been agreed with Clement (the maintainer of the Quarkus messaging extensions). He suggested to use the emitter.hasRequests method to be sure the broker can handle more messages. The hasRequests works like a capacity barrier: when we send up to 1024 messages (1024 is the default value) without getting an ack from the broker, we wait until sending more messages).

Update: there seems to be two upper limits, (1) the one configured by the `max-inflight-messages` which is the one used by the `hasRequests` method and default value is 1024; and (2) the buffer limit which is used by the overflow strategy and default value is 128. Therefore, for this solution to work, we need to set the `max-inflight-messages` limit to the same as the buffer limit. 

Other solutions that we spoke about:
- use `@Outgoing` annotation along with the `@Blocking` annotation: this would hit a very performance penalty
- use `@OnOverflow(unbounded buffer)`: there is a OOM risk here that we should avoid
- use  `mp.messaging.outgoing.your-kafka-channel.waitForWriteCompletion=false`: but we can't ensure the messages are sent
- use `mp.messaging.outgoing.your-kafka-channel.max-inflight-messages=0`: again, we can't guarantee that the messages are sent

## Testing

Since I could not reproduce the "SRMSG00034: Insufficient downstream requests to emit item" issue, I asked Shubham for running the performance test against this pull request.  

## Testing locally

1.- podman-compose up
2.- Mock the prometheus server at localhost:

Create the stub directory:
`mkdir -p stub/__files`

Add a stub file to return some data: `stub/__files/swatch-1805.json`

```
cat > stub/__files/swatch-1805.json <<EOF
{
    "status" : "success",
    "data" : {
      "resultType" : "matrix",
      "result" : [
        {
          "metric" : {
            "_id" : "cluster id",
            "support" : "sla",
            "usage" : "usage",
            "product" : "product",
            "billing_marketplace" : "billing_marketplace",
            "billing_marketplace_account" : "billing_marketplace_account",
            "ebs_account" : "account"
          },
          "values": [[ $(date +%s), "1" ]]
        }
      ]
    }
}
EOF
```

Add a stub file to return no data: `stub/__files/empty.json`

```
cat > stub/__files/empty.json <<EOF
{
    "status" : "success",
    "data" : {
      "resultType" : "matrix",
      "result" : []
    }
}
EOF
```

Run wiremock:
`podman run -it --rm -p 8101:8080 --name wiremock -v $PWD/stub:/home/wiremock:z wiremock/wiremock:2.32.0 --verbose`

Configure stubbing for prometheus:

- this is to return some data when asking for instance-hours:
`curl -X POST --data '{ "priority": 1, "request": { "urlPath": "/api/v1/query_range", "method": "GET", "queryParameters": {"query": {"contains":"instance_hours"}}}, "response": { "status": 200, "bodyFileName": "swatch-1805.json", "headers":{"Content-Type":"application/json"} }}' http://localhost:8101/__admin/mappings/new`

- this is to return no data otherwise:
`curl -X POST --data '{ "priority": 5, "request": { "urlPath": "/api/v1/query_range", "method": "GET" }, "response": { "status": 200, "bodyFileName": "empty.json", "headers":{"Content-Type":"application/json"} }}' http://localhost:8101/__admin/mappings/new`

3.- Start the swatch metrics app: `PROM_URL=http://localhost:8101/api/v1 ENABLE_SYNCHRONOUS_OPERATIONS=true ./gradlew :swatch-metrics:quarkusDev`

4.- Trigger the metrics from prometheus:
```
http POST :8000/api/swatch-metrics/v1/internal/metering/rosa?orgId=16790890 \
x-rh-swatch-synchronous-request:true
```

5.- Confirm the events have been produced in the Kafka topic "platform.rhsm-subscriptions.service-instance-ingress".

You should see two events in http://localhost:3030/#/cluster/default/topic/n/platform.rhsm-subscriptions.service-instance-ingress/

### CODE REVIEW

**Comment by Sgitario** [comment]
Adding the QE label to wait for Shubham's feedback and confirm that the issue is gone.

**Comment by shubham-html-css-js** [comment]
I tested the recent changes and I was able to send 50k events.

**Review by mstead — CHANGES_REQUESTED** [review]
Please add the details of the code changes to the actual git commit message so that we do not have to rely on the JIRA ticket.

Otherwise this looks good to me!

**Comment by Sgitario** [comment]
> Please add the details of the code changes to the actual git commit message so that we do not have to rely on the JIRA ticket.
> 
> Otherwise this looks good to me!

what details do you mean? All the information is part of the pull request that once merged, will be included in the merge commit.

**Comment by Sgitario** [comment]
@mstead PR updated

**Review by mstead — APPROVED** [review]

**Pull request merged**
